### PR TITLE
Implementing a setting cache

### PR DESCRIPTION
**Pull request by MohamedBassem**
Implementing a setting cache

ghstack-source-id: c906f41fd517a0021667280bf1bed438bf93e9e2
Pull Request resolved: https://github.com/devtari-io/cronback/pull/19

### CODE REVIEW

File: cronback-lib/project_settings.rs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+use crate::types::{ProjectId, RequestId};
+use crate::GrpcClientFactory;
+
+type MetadataClientFactory = Arc<
+    Box<dyn GrpcClientFactory<ClientType = ScopedMetadataSvcClient> + Send>,
+>;
+
+#[derive(Error, Debug, Clone)]
+pub enum ProjectSettingError {
+    // Optimally this should wrap a `GrpcClientError` but unfortauntly, it's
```
Comment (AhmedSoliman) [inline cronback-lib/project_settings.rs:20]:
```suggestion
    // Optimally this should wrap a `GrpcClientError` but unfortunately, it's
```

File: cronback-lib/project_settings.rs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    #[error("failed to create metadata svc client: {0}")]
+    Client(String),
+    #[error("setting grpc called failed with code {0}")]
+    Server(#[from] tonic::Status),
+}
+
+/// A TTL-based read-through async project setting cache. A single instance of
+/// this struct caches only a single setting type.
+///
+/// Reading from the cache will return the value immedietly if it exists, and
```
Comment (AhmedSoliman) [inline cronback-lib/project_settings.rs:32]:
```suggestion
/// Reading from the cache will return the value immediately if it exists, and
```

File: cronback-lib/project_settings.rs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+/// will fetch it if it doesn't exist. Notes:
+/// 1. If multiple concurrent callers request the same uncached value, only one
+///    of them will fetch it and the rest will wait for the result.
+/// 2. If the cached value exceeded the TTL, it will get refetched on the next
+///    read.
+/// 3. If the fetch from source fails, waiting callers will get fullfilled with
```
Comment (AhmedSoliman) [inline cronback-lib/project_settings.rs:38]:
```suggestion
/// 3. If the fetch from source fails, waiting callers will get fulfilled with
```

File: cronback-lib/Cargo.toml [hunk 4]
```diff
@@ -65,3 +65,5 @@ sea-orm-migration = { version = "0.11.0", features = [
   "sqlx-sqlite",              # `DATABASE_DRIVER` feature
   "sqlx-postgres",            # `DATABASE_DRIVER` feature
 ] }
+
+moka = { version = "0.11.2", features = ["future"] }
```
Comment (AhmedSoliman) [inline cronback-lib/Cargo.toml:69]: As you mentioned offline, moka is a big dependency and it sounds like an overkill for our simple use-case. That said, I don't mind leaving it to get a feel and learn. We eventually should standardise all internal caches on a single library whenever possible.

File: cronback-lib/project_settings.rs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+///    the failed Result. However, the value will get refetched on the next read
+///    attempt (even if it didn't exceed the TTL).
+/// 4. Cache evictions happen in the background in a separate thread pool.
+pub struct ProjectSetting<V, F> {
+    client_factory: MetadataClientFactory,
+    fetcher: F,
+    cache: Cache<ValidShardedId<ProjectId>, Result<V, ProjectSettingError>>,
```
Comment (AhmedSoliman) [inline cronback-lib/project_settings.rs:45]: I'm a bit torn on whether the key should be a decorated `ValidShardedId` or not. I haven't formed a strong opinion about where to enforce ValidShardedId, I'm comfortable leaving this as is, and maybe it's something we should reflect on later and see how we both feel about it.

**Review by AhmedSoliman — COMMENTED** [review]
A few comments and thoughts but generally looks peaceful.

File: cronback-lib/project_settings.rs (hunk 5 above)
Comment (MohamedBassem) [reply cronback-lib/project_settings.rs:45]: I like the idea of caching the non-decorated id, will switch to that.

File: cronback-lib/project_settings.rs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+/// will fetch it if it doesn't exist. Notes:
+/// 1. If multiple concurrent callers request the same uncached value, only one
+///    of them will fetch it and the rest will wait for the result.
+/// 2. If the cached value exceeded the TTL, it will get refetched on the next
+///    read.
+/// 3. If the fetch from source fails, waiting callers will get fullfilled with
+///    the failed Result. However, the value will get refetched on the next read
+///    attempt (even if it didn't exceed the TTL).
+/// 4. Cache evictions happen in the background in a separate thread pool.
+pub struct ProjectSetting<V, F> {
```
[the remark this answers is not in the document]
Comment (MohamedBassem) [reply cronback-lib/project_settings.rs:42]: > The current name implies that users can get an individual setting for a given project.

It is actually exactly that. As of now, there's no entity called "project settings". My idea was to keep them disaggregated such that if you want to get only the project status and not the project notification settings you can do that. That's why it's generic over the value. The value here will be the piece of the setting that you'll fetch from the project, and the fetcher will be the lambda that gets a metadata client and calls the corresponding metadata API on it.

```
type NotificationSettings = ProjectSetting<ProjectNotificationSetting, Fn(&ProjectId, &MetadataClient) -> ProjectNotificationSetting;

type ProjectStatus = ProjectSetting<ProjectStatus, Fn(&ProjectId, &MetadataClient) -> ProjectStatus;
```

Hope that makes sense.

[the remark this answers is not in the document]
Comment (AhmedSoliman) [reply cronback-lib/project_settings.rs:42]: Interesting approach.

**Review by AhmedSoliman — DISMISSED** [review]

**Pull request merged**